# Abhishek Sharma 

## CS Second Year ; Section : "I" ; Roll no.: 01 

## Enrollment No.: 12019009001127 

### Advanced AI/ML Lab 1 - 02.02.2021 

Importing the library : ```numpy```

In [1]:
import numpy as np

In [2]:
a = np.array([1,2,3]) #initialize the array using the numpy module
print (a) #print the array
print (type(a)) #data type of the variable
print (a.ndim) # return the dimension of the array
print (a.shape) # shape of the array
print (len(a)) # length of the array

[1 2 3]
<class 'numpy.ndarray'>
1
(3,)
3


In [4]:
# generate one dimensional array using numpy
np.arange(15)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

In [5]:
# matrix like sructure or reshape function
np.arange(15).reshape(3,5)

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

In [6]:
np.linspace(1,8,4) 
# four equi-distant points in b/w 1 to 8

array([1.        , 3.33333333, 5.66666667, 8.        ])

In [7]:
np.linspace(1,8,4, endpoint = False) 
# exclude the end point

array([1.  , 2.75, 4.5 , 6.25])

In [8]:
np.random.randint(1,100,6) 
# random six elements in b/w the range 1 to 100

array([85, 94, 80, 53, 40,  5])

In [9]:
np.random.rand(4)
# taking 4 random values between 0 to 1

array([0.12129421, 0.57849506, 0.14199295, 0.00438752])

In [10]:
np.eye(3) 
# identity matrix of 3x3

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [11]:
np.eye(5)
# identity matrix 5x5

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

In [12]:
np.zeros(3)
# zero arrays of dimension 3

array([0., 0., 0.])

In [13]:
np.random.randn(3,2) 
# 3x2 matrix random variable matrix

array([[ 0.70114703, -0.02236597],
       [ 0.02923078, -0.07167346],
       [-0.32627708, -0.26481326]])

In [14]:
np.empty((3,3))

#empty matrix of 3x3

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [18]:
np.arange(0,2,0.3)

# print the numbers in b/w 0 to 2 with space of 0.3

array([0. , 0.3, 0.6, 0.9, 1.2, 1.5, 1.8])

In [16]:
np.ones((2,3,4), dtype = np.int16)

array([[[1, 1, 1, 1],
        [1, 1, 1, 1],
        [1, 1, 1, 1]],

       [[1, 1, 1, 1],
        [1, 1, 1, 1],
        [1, 1, 1, 1]]], dtype=int16)

In [17]:
np.array([[1,2],[3,4]], dtype = complex)

# Complex matrix

array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]])

In [19]:
A = np.array([[2,1],[0,1]])
B = np.array([[2,0],[3,4]])

# initializing two matrices

In [20]:
A + B
# addition of the matrices

array([[4, 1],
       [3, 5]])

In [21]:
np.add(A,B)
# adding using numpy

array([[4, 1],
       [3, 5]])

In [23]:
A * B
# matrix multiplication but it is element wise multiplication

array([[4, 0],
       [0, 4]])

In [24]:
A @ B
# normal matrix multiplication

array([[7, 4],
       [3, 4]])

In [25]:
A.dot(B)
# normal matrix multiplication or the dot product

array([[7, 4],
       [3, 4]])

In [27]:
B.T
# Transpose of B Matrix

array([[2, 3],
       [0, 4]])

In [30]:
A.flatten()
# turn into one dimensional matrix in a row wise manner or flattening the matrix

array([2, 1, 0, 1])

In [29]:
B<3

# element wise condition checking of the B matrix

array([[ True,  True],
       [False, False]])

In [31]:
A.sum() # sum of all the elements

4

In [32]:
A.sum(axis=0) # column wise element addition of the A matrix

array([2, 2])

In [33]:
A.sum(axis=1) # row wise addition of A matrix

array([3, 1])

In [34]:
A

array([[2, 1],
       [0, 1]])

In [35]:
A.cumsum(axis=1) # cummulative sum of the elements row wise

array([[2, 3],
       [0, 1]], dtype=int32)

In [38]:
A.min()
# minimum element of the matrix

0

In [39]:
A.max()
# maximum element of the matrix

2

In [40]:
# New segment

a = np.arange(4)**3

In [41]:
print (a)

[ 0  1  8 27]


In [42]:
a[2] #2nd index position of the array

8

In [43]:
a[::-1] # reverse the array

array([27,  8,  1,  0], dtype=int32)

In [44]:
a[0:4:1] # print the elements of the array with 1 position index

array([ 0,  1,  8, 27], dtype=int32)

In [45]:
a[1,...] # a[1] element

array(1, dtype=int32)

In [46]:
a[2,...] #a[2] element

array(8, dtype=int32)

In [47]:
a[a>5] # check elements greater than 5 in the array 'a'

array([ 8, 27], dtype=int32)

In [48]:
X = a[0:4]
print (X)

# taking all the elements

[ 0  1  8 27]


In [51]:
X[:]=99 # replace all the elements of X with 99
print (X)

[99 99 99 99]


### DataFrame initialization using the ```pandas``` library 

In [52]:
import pandas as pd

In [53]:
arr = np.array([1,3,5,7,9])
s2 = pd.Series(arr) # creating series from the array
print (s2)
print (type(s2))

0    1
1    3
2    5
3    7
4    9
dtype: int32
<class 'pandas.core.series.Series'>


In [55]:
# creating the series 
s2 = pd.Series([10,20,30])
print (s2)

# Creating the Data Frame from the matrix data using pandas
s3 = pd.DataFrame([[1,2],[3,4]],columns=['A','B'], index = ['C','D'])
print (s3)
print (type(s3))

0    10
1    20
2    30
dtype: int64
   A  B
C  1  2
D  3  4
<class 'pandas.core.frame.DataFrame'>


In [56]:
s4 = pd.DataFrame(np.random.randn(20,7), columns = ['a','b','c','d','e','f','g'])
print (s4)

# providing column names to the dataframe

           a         b         c         d         e         f         g
0  -0.923618  1.881778 -0.389276  0.775150  1.915383 -0.007260 -1.134208
1  -0.538072 -0.080831 -0.796555 -0.278779  0.069101  0.888007  0.012214
2   2.180868 -1.747470  0.781028 -0.369416 -0.072360  0.548173  0.047412
3   1.116868 -0.504462  0.873826 -1.260247 -0.746652 -0.232982 -0.588743
4  -0.277801 -1.123292  0.256486  1.036584 -0.301103 -1.128012  0.566962
5   1.082634 -0.447155 -0.551533  0.933202  1.440657  1.680650  0.165088
6   0.526940 -0.323593 -2.045275 -1.101051 -0.135043 -0.267078 -1.796928
7   1.193797  0.187204  0.703582 -0.227264  0.531728 -0.295992  0.182776
8  -0.104294  0.525912  1.351340 -0.547446  0.048356 -0.170555  0.642638
9   0.033289  0.465015  0.386574  1.264157 -0.087108  0.630022 -0.446349
10 -0.492799  0.883487  0.078328  0.963538 -0.042213 -0.878139 -0.543125
11 -1.925714 -0.479302  0.355445 -0.492763 -1.438186 -2.047101 -0.186275
12 -0.111234  1.510605  0.061694 -0.207015 -0.67595

In [57]:
s4[(5<s4.index) &(s4.index<10)]

# checking the certain conditions

a         b         c         d         e         f         g
6  0.526940 -0.323593 -2.045275 -1.101051 -0.135043 -0.267078 -1.796928
7  1.193797  0.187204  0.703582 -0.227264  0.531728 -0.295992  0.182776
8 -0.104294  0.525912  1.351340 -0.547446  0.048356 -0.170555  0.642638
9  0.033289  0.465015  0.386574  1.264157 -0.087108  0.630022 -0.446349

In [58]:
s4.head()

a         b         c         d         e         f         g
0 -0.923618  1.881778 -0.389276  0.775150  1.915383 -0.007260 -1.134208
1 -0.538072 -0.080831 -0.796555 -0.278779  0.069101  0.888007  0.012214
2  2.180868 -1.747470  0.781028 -0.369416 -0.072360  0.548173  0.047412
3  1.116868 -0.504462  0.873826 -1.260247 -0.746652 -0.232982 -0.588743
4 -0.277801 -1.123292  0.256486  1.036584 -0.301103 -1.128012  0.566962

In [59]:
s4.tail()

a         b         c         d         e         f         g
15  1.437014 -0.623354 -0.016936 -0.414716 -0.611414 -1.149088 -0.360014
16 -2.706737  0.436487  1.007067  2.003439  1.091563 -0.421391  0.656348
17 -0.326253 -0.578320  0.656961 -0.918057 -0.150000 -0.273481 -0.705258
18  0.110080  0.165280  1.656131  0.788284 -0.497398 -2.265728  1.647139
19 -0.697831  1.401574  0.479545  1.554887 -0.184202  0.500018  0.440924

In [60]:
s4.describe()

# description of the data frame

a          b          c          d          e          f  \
count  20.000000  20.000000  20.000000  20.000000  20.000000  20.000000   
mean   -0.006671   0.052526   0.119298   0.170618  -0.037549  -0.196512   
std     1.141324   0.950694   1.057746   0.925727   0.873191   1.088001   
min    -2.706737  -1.747470  -2.775911  -1.260247  -1.676785  -2.265728   
25%    -0.504118  -0.522927  -0.110021  -0.434228  -0.525902  -0.940607   
50%    -0.107764   0.042224   0.334443  -0.217139  -0.111076  -0.250030   
75%     0.802798   0.579267   0.722944   0.940786   0.184758   0.566383   
max     2.180868   1.881778   1.656131   2.003439   1.915383   1.762622   

               g  
count  20.000000  
mean   -0.152298  
std     0.776661  
min    -1.796928  
25%    -0.554530  
50%    -0.107634  
75%     0.247313  
max     1.647139

In [61]:
s4[['b','e']]

# show particular column in the form of Data Frame

b         e
0   1.881778  1.915383
1  -0.080831  0.069101
2  -1.747470 -0.072360
3  -0.504462 -0.746652
4  -1.123292 -0.301103
5  -0.447155  1.440657
6  -0.323593 -0.135043
7   0.187204  0.531728
8   0.525912  0.048356
9   0.465015 -0.087108
10  0.883487 -0.042213
11 -0.479302 -1.438186
12  1.510605 -0.675954
13  0.739333 -1.676785
14 -1.238382  0.770649
15 -0.623354 -0.611414
16  0.436487  1.091563
17 -0.578320 -0.150000
18  0.165280 -0.497398
19  1.401574 -0.184202

In [62]:
s4[0:3]

# slicing operation using the : operator in the Dataframe

a         b         c         d         e         f         g
0 -0.923618  1.881778 -0.389276  0.775150  1.915383 -0.007260 -1.134208
1 -0.538072 -0.080831 -0.796555 -0.278779  0.069101  0.888007  0.012214
2  2.180868 -1.747470  0.781028 -0.369416 -0.072360  0.548173  0.047412

In [64]:
s4.iloc[0:3]

# index location and slicing operation in the Dataframe

a         b         c         d         e         f         g
0 -0.923618  1.881778 -0.389276  0.775150  1.915383 -0.007260 -1.134208
1 -0.538072 -0.080831 -0.796555 -0.278779  0.069101  0.888007  0.012214
2  2.180868 -1.747470  0.781028 -0.369416 -0.072360  0.548173  0.047412

In [65]:
s4.loc[[2,3],['a','b']]

# specify the particular elements in a Data Frame

a         b
2  2.180868 -1.747470
3  1.116868 -0.504462

In [66]:
s4[2<s4]

# checking which elements are less than 2

a   b   c         d   e   f   g
0        NaN NaN NaN       NaN NaN NaN NaN
1        NaN NaN NaN       NaN NaN NaN NaN
2   2.180868 NaN NaN       NaN NaN NaN NaN
3        NaN NaN NaN       NaN NaN NaN NaN
4        NaN NaN NaN       NaN NaN NaN NaN
5        NaN NaN NaN       NaN NaN NaN NaN
6        NaN NaN NaN       NaN NaN NaN NaN
7        NaN NaN NaN       NaN NaN NaN NaN
8        NaN NaN NaN       NaN NaN NaN NaN
9        NaN NaN NaN       NaN NaN NaN NaN
10       NaN NaN NaN       NaN NaN NaN NaN
11       NaN NaN NaN       NaN NaN NaN NaN
12       NaN NaN NaN       NaN NaN NaN NaN
13       NaN NaN NaN       NaN NaN NaN NaN
14       NaN NaN NaN       NaN NaN NaN NaN
15       NaN NaN NaN       NaN NaN NaN NaN
16       NaN NaN NaN  2.003439 NaN NaN NaN
17       NaN NaN NaN       NaN NaN NaN NaN
18       NaN NaN NaN       NaN NaN NaN NaN
19       NaN NaN NaN       NaN NaN NaN NaN

In [67]:
s4.mean()

# mean of each column

a   -0.006671
b    0.052526
c    0.119298
d    0.170618
e   -0.037549
f   -0.196512
g   -0.152298
dtype: float64

In [68]:
s4.mean(1)

# mean of each row

0     0.302564
1    -0.103559
2     0.195462
3    -0.191770
4    -0.138597
5     0.614792
6    -0.734576
7     0.325119
8     0.249422
9     0.320800
10   -0.004418
11   -0.887699
12    0.330247
13   -0.504609
14   -0.365225
15   -0.248358
16    0.295254
17   -0.327773
18    0.229112
19    0.499273
dtype: float64

In [69]:
s4.drop(5, axis = 0)

# dropping the index 5 row from the data frame

a         b         c         d         e         f         g
0  -0.923618  1.881778 -0.389276  0.775150  1.915383 -0.007260 -1.134208
1  -0.538072 -0.080831 -0.796555 -0.278779  0.069101  0.888007  0.012214
2   2.180868 -1.747470  0.781028 -0.369416 -0.072360  0.548173  0.047412
3   1.116868 -0.504462  0.873826 -1.260247 -0.746652 -0.232982 -0.588743
4  -0.277801 -1.123292  0.256486  1.036584 -0.301103 -1.128012  0.566962
6   0.526940 -0.323593 -2.045275 -1.101051 -0.135043 -0.267078 -1.796928
7   1.193797  0.187204  0.703582 -0.227264  0.531728 -0.295992  0.182776
8  -0.104294  0.525912  1.351340 -0.547446  0.048356 -0.170555  0.642638
9   0.033289  0.465015  0.386574  1.264157 -0.087108  0.630022 -0.446349
10 -0.492799  0.883487  0.078328  0.963538 -0.042213 -0.878139 -0.543125
11 -1.925714 -0.479302  0.355445 -0.492763 -1.438186 -2.047101 -0.186275
12 -0.111234  1.510605  0.061694 -0.207015 -0.675954  1.762622 -0.028992
13  0.709519  0.739333 -2.775911  0.139961 -1.676785  0.621013 -1.289395
14 -0.420069 -1.238382  0.313440 -0.230092  0.770649 -1.423941 -0.328181
15  1.437014 -0.623354 -0.016936 -0.414716 -0.611414 -1.149088 -0.360014
16 -2.706737  0.436487  1.007067  2.003439  1.091563 -0.421391  0.656348
17 -0.326253 -0.578320  0.656961 -0.918057 -0.150000 -0.273481 -0.705258
18  0.110080  0.165280  1.656131  0.788284 -0.497398 -2.265728  1.647139
19 -0.697831  1.401574  0.479545  1.554887 -0.184202  0.500018  0.440924

In [70]:
s4.drop('d', axis = 1)
# droping the 'd' column

a         b         c         e         f         g
0  -0.923618  1.881778 -0.389276  1.915383 -0.007260 -1.134208
1  -0.538072 -0.080831 -0.796555  0.069101  0.888007  0.012214
2   2.180868 -1.747470  0.781028 -0.072360  0.548173  0.047412
3   1.116868 -0.504462  0.873826 -0.746652 -0.232982 -0.588743
4  -0.277801 -1.123292  0.256486 -0.301103 -1.128012  0.566962
5   1.082634 -0.447155 -0.551533  1.440657  1.680650  0.165088
6   0.526940 -0.323593 -2.045275 -0.135043 -0.267078 -1.796928
7   1.193797  0.187204  0.703582  0.531728 -0.295992  0.182776
8  -0.104294  0.525912  1.351340  0.048356 -0.170555  0.642638
9   0.033289  0.465015  0.386574 -0.087108  0.630022 -0.446349
10 -0.492799  0.883487  0.078328 -0.042213 -0.878139 -0.543125
11 -1.925714 -0.479302  0.355445 -1.438186 -2.047101 -0.186275
12 -0.111234  1.510605  0.061694 -0.675954  1.762622 -0.028992
13  0.709519  0.739333 -2.775911 -1.676785  0.621013 -1.289395
14 -0.420069 -1.238382  0.313440  0.770649 -1.423941 -0.328181
15  1.437014 -0.623354 -0.016936 -0.611414 -1.149088 -0.360014
16 -2.706737  0.436487  1.007067  1.091563 -0.421391  0.656348
17 -0.326253 -0.578320  0.656961 -0.150000 -0.273481 -0.705258
18  0.110080  0.165280  1.656131 -0.497398 -2.265728  1.647139
19 -0.697831  1.401574  0.479545 -0.184202  0.500018  0.440924

In [74]:
s4.drop('a',axis = 1, inplace = True)
print (s4)

# dropping a column

           b         c         e         f         g
0   1.881778 -0.389276  1.915383 -0.007260 -1.134208
1  -0.080831 -0.796555  0.069101  0.888007  0.012214
2  -1.747470  0.781028 -0.072360  0.548173  0.047412
3  -0.504462  0.873826 -0.746652 -0.232982 -0.588743
4  -1.123292  0.256486 -0.301103 -1.128012  0.566962
5  -0.447155 -0.551533  1.440657  1.680650  0.165088
6  -0.323593 -2.045275 -0.135043 -0.267078 -1.796928
7   0.187204  0.703582  0.531728 -0.295992  0.182776
8   0.525912  1.351340  0.048356 -0.170555  0.642638
9   0.465015  0.386574 -0.087108  0.630022 -0.446349
10  0.883487  0.078328 -0.042213 -0.878139 -0.543125
11 -0.479302  0.355445 -1.438186 -2.047101 -0.186275
12  1.510605  0.061694 -0.675954  1.762622 -0.028992
13  0.739333 -2.775911 -1.676785  0.621013 -1.289395
14 -1.238382  0.313440  0.770649 -1.423941 -0.328181
15 -0.623354 -0.016936 -0.611414 -1.149088 -0.360014
16  0.436487  1.007067  1.091563 -0.421391  0.656348
17 -0.578320  0.656961 -0.150000 -0.273481 -0.

In [75]:
s4['h']=np.random.rand(20,1)
print (s4) 

#adding a new column with same data frame

           b         c         e         f         g         h
0   1.881778 -0.389276  1.915383 -0.007260 -1.134208  0.185472
1  -0.080831 -0.796555  0.069101  0.888007  0.012214  0.957329
2  -1.747470  0.781028 -0.072360  0.548173  0.047412  0.048276
3  -0.504462  0.873826 -0.746652 -0.232982 -0.588743  0.221025
4  -1.123292  0.256486 -0.301103 -1.128012  0.566962  0.965142
5  -0.447155 -0.551533  1.440657  1.680650  0.165088  0.906482
6  -0.323593 -2.045275 -0.135043 -0.267078 -1.796928  0.696071
7   0.187204  0.703582  0.531728 -0.295992  0.182776  0.340151
8   0.525912  1.351340  0.048356 -0.170555  0.642638  0.263669
9   0.465015  0.386574 -0.087108  0.630022 -0.446349  0.014660
10  0.883487  0.078328 -0.042213 -0.878139 -0.543125  0.793390
11 -0.479302  0.355445 -1.438186 -2.047101 -0.186275  0.637648
12  1.510605  0.061694 -0.675954  1.762622 -0.028992  0.591581
13  0.739333 -2.775911 -1.676785  0.621013 -1.289395  0.265380
14 -1.238382  0.313440  0.770649 -1.423941 -0.328181  0

***************************************************

### Linear Regression Model 

Package = ```sklearn```

dataset = **Boston Dataset**

In [76]:
from sklearn.datasets import load_boston

In [77]:
boston = load_boston()

In [79]:
type(boston)

sklearn.utils.Bunch

In [81]:
print (boston.DESCR)  # descripetion of the dataset

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [82]:
boston.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [83]:
boston.data

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [85]:
type(boston.feature_names)

numpy.ndarray

In [86]:
boston.data.shape

# shape of the dataset

(506, 13)

In [90]:
data = pd.DataFrame(boston.data, columns = boston.feature_names)

# Creating the data frame using the pandas function

In [91]:
data.head(5)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [93]:
data['MEDV'] = pd.DataFrame(boston.target) 

# adding the target values in the dataframe and named it as MEDV

In [94]:
data.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [95]:
# checking the corealation

pd.DataFrame (data.corr().round(2))

CRIM    ZN  INDUS  CHAS   NOX    RM   AGE   DIS   RAD   TAX  PTRATIO  \
CRIM     1.00 -0.20   0.41 -0.06  0.42 -0.22  0.35 -0.38  0.63  0.58     0.29   
ZN      -0.20  1.00  -0.53 -0.04 -0.52  0.31 -0.57  0.66 -0.31 -0.31    -0.39   
INDUS    0.41 -0.53   1.00  0.06  0.76 -0.39  0.64 -0.71  0.60  0.72     0.38   
CHAS    -0.06 -0.04   0.06  1.00  0.09  0.09  0.09 -0.10 -0.01 -0.04    -0.12   
NOX      0.42 -0.52   0.76  0.09  1.00 -0.30  0.73 -0.77  0.61  0.67     0.19   
RM      -0.22  0.31  -0.39  0.09 -0.30  1.00 -0.24  0.21 -0.21 -0.29    -0.36   
AGE      0.35 -0.57   0.64  0.09  0.73 -0.24  1.00 -0.75  0.46  0.51     0.26   
DIS     -0.38  0.66  -0.71 -0.10 -0.77  0.21 -0.75  1.00 -0.49 -0.53    -0.23   
RAD      0.63 -0.31   0.60 -0.01  0.61 -0.21  0.46 -0.49  1.00  0.91     0.46   
TAX      0.58 -0.31   0.72 -0.04  0.67 -0.29  0.51 -0.53  0.91  1.00     0.46   
PTRATIO  0.29 -0.39   0.38 -0.12  0.19 -0.36  0.26 -0.23  0.46  0.46     1.00   
B       -0.39  0.18  -0.36  0.05 -0.38  0.13 -0.27  0.29 -0.44 -0.44    -0.18   
LSTAT    0.46 -0.41   0.60 -0.05  0.59 -0.61  0.60 -0.50  0.49  0.54     0.37   
MEDV    -0.39  0.36  -0.48  0.18 -0.43  0.70 -0.38  0.25 -0.38 -0.47    -0.51   

            B  LSTAT  MEDV  
CRIM    -0.39   0.46 -0.39  
ZN       0.18  -0.41  0.36  
INDUS   -0.36   0.60 -0.48  
CHAS     0.05  -0.05  0.18  
NOX     -0.38   0.59 -0.43  
RM       0.13  -0.61  0.70  
AGE     -0.27   0.60 -0.38  
DIS      0.29  -0.50  0.25  
RAD     -0.44   0.49 -0.38  
TAX     -0.44   0.54 -0.47  
PTRATIO -0.18   0.37 -0.51  
B        1.00  -0.37  0.33  
LSTAT   -0.37   1.00 -0.74  
MEDV     0.33  -0.74  1.00

RM and MEDV is having the highest corelation and hence RM is the variable with which we can train the dataset and check the target values as MEDV

In [96]:
x = data['RM']

In [97]:
y = data['MEDV']

In [99]:
type(x)

# both the RM and MEDV is fetched as the series object

pandas.core.series.Series

In [100]:
pd.DataFrame ([x,y]).transpose().head()

# transpose is used here for better visualization and benefit for the future use

RM  MEDV
0  6.575  24.0
1  6.421  21.6
2  7.185  34.7
3  6.998  33.4
4  7.147  36.2

Importing the ```LinearRegression``` Algorithm/Model from the ```sklearn.linear_model``` to create the linear regression model for the dataset

In [102]:
from sklearn.linear_model import LinearRegression

In [103]:
model1 = LinearRegression()

In [104]:
x = pd.DataFrame(x)
y = pd.DataFrame(y)

In [106]:
x.head() # creating the dataframe for fitting the model in the Linear regression

RM
0  6.575
1  6.421
2  7.185
3  6.998
4  7.147

In [107]:
y.head()

MEDV
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2

In [108]:
model1.fit(x,y) # fitting the model with the data frames of x and y

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [112]:
x_test = [6.41,6.66,7.5]
x_new = pd.DataFrame(x_test)

In [114]:
y_pred = model1.predict(x_new)

In [117]:
y_pred = pd.DataFrame(y_pred)

In [118]:
y_pred

# predicting the values in the dataframe format

0
0  23.673898
1  25.949425
2  33.595197

### Result : Predicting the values as per the test dataset

**Here's the final table**

|Test Value 'RM'|Predicted Value 'MEDV'|
|:---:|:---:|
|6.41|23.67|
|6.66|25.95|
|7.5|33.60|

### Thank You!

*****************************